### DISCUSSION TITLE
Why does it seem that setting component properties with a ternary using rx.Base properies doesn't seem to work for me.

### DISCUSSION THREAD

**Discussion by nevdelap** [reflex-dev/reflex#1644] [Q&A] [answered]
Hi there. I'm having an issue with rx.Base properties and setting component properties with conditionals based on them. This is minimal example to show an issue.

### Setup

Python: 3.11.4
Reflex: 0.2.4

```bash
mkdir issue
cd issue
reflex init
# copy the code below into issue/issue.py
reflex run
```

### Code

```
import reflex as rx


class Thing(rx.Base):
    description: str
    red: bool


class State(rx.State):
    things: list[Thing] = [
        Thing(
            description=(
                "should be red, shows ternary on rx.Base property seeming to work "
                + "but maybe not really"
            ),
            red=True,
        ),
        Thing(
            description=(
                "should be blue, shows ternary on rx.Base property doing something "
                + "unexpected"
            ),
            red=False,
        ),
    ]


def thing_text(thing: Thing):
    return rx.vstack(
        rx.text(
            thing.description,
            color="red" if thing.red else "blue"
        )
    )


def index() -> rx.Component:
    return rx.vstack(
        rx.foreach(
            State.things,
            thing_text,
        ),
        rx.text(
            "should be red, shows ternary working as expected",
            color="red" if "1" == "1" else "blue"
        ),
        rx.text(
            "should be blue, shows ternary working as expected",
            color="red" if "1" == "2" else "blue"
        ),
    )


app = rx.App(state=State)
app.add_page(index)
app.compile()

```

### The Unexpected Result

The line highlighted in yellow shows the issue. It should be blue, and it seems to be something to do with using state properties in ternary conditions. In my app the real thing that is going on is I'm trying to do things like `display="" if xxx else "none"` where xxx is some conditional expression using an rx.Base derived class.

![image](https://github.com/reflex-dev/reflex/assets/583807/ba3d5040-3646-4ccb-b857-a0691303f885)

### What's Going On?
My suspicion is that it is a bug, but just as likely is there is some limitation that I'm not aware of, or I'm misunderstanding something about how and where client side things can be used.

I did also have issues when doing conditionals with `is None` and `is not None` that were not working and I thought that must not be supported, but maybe it is and it's the same issue.

**Comment by nevdelap** [comment]
I checked that the behaviour is the same in 0.2.5. It is.

**Comment by masenf** [comment]
@nevdelap the problem here is that the frontend is _compiled_ into javascript code. So those python ternary expressions are always evaluating to `True` at compile time, since the Base Vars are truthy.

An unfortunate wrinkle in the API requires the use of `rx.cond` to compile ternary expressions for rendering on the frontend. Similar to why `rx.foreach` must be used, instead of native python `for` loop.

```python
def thing_text(thing: Thing):
    return rx.vstack(
        rx.text(
            thing.description,
            color=rx.cond(thing.red, "red", "blue")
        )
    )
```

If you have any ideas for how we might compile these python native control structures into javascript, that would definitely be ideal.

I have been thinking about attempting to parse the AST of the render function at compile time and systematically replace `if` and `for` constructs with their appropriate counterparts, but it's not straightforward and with other function calls it gets hairy fast (and other lower hanging fruit is continuing to take precedent).

**Comment by nevdelap** [comment]
Thanks @masenf. I thought is was something like that (in the Javascript part of things.) I'll have a closer look.

**Comment by nevdelap** [comment]
rx.cond is nice. I was trying to set the "display" css property from state, and now I see how I should be doing it from the example in the Cond doco. I think I didn't find it because I wasn't thinking about it as being to do with Layout.

**Comment by nevdelap** [comment]
Hey again. I've been trying to use rx.cond and it appears that it doesn't react to the condition changing. I've gotten as far as demonstrating the issue, not as far as digging into rx.cond to understand anything in the implementation so far.

This is a minimal example:

```
# pylint: disable=missing-class-docstring
# pylint: disable=missing-function-docstring
# pylint: disable=missing-module-docstring
# pylint: disable=redefined-outer-name

import reflex as rx


class Thing(rx.Base):
    value: str
    new_value: str | None
    is_editing: bool  # new_value is not None could serve this purpose, but to keep things clear.


class State(rx.State):
    things: list[Thing] = [
        Thing(
            value ="Hello from this example. 👋",
            new_value = None,
            is_editing=False
        ),
        Thing(
            value="The **other** thing.",
            new_value = None,
            is_editing=False
        ),
        Thing(
            value="And its friend.",
            new_value = None,
            is_editing=False
        ),
    ]

    def edit(self, index: int) -> None:
        self.things[index].new_value = self.things[index].value
        self.things[index].is_editing = True

    def update_new_value(self, index:int, value: str) -> None:
        self.things[index].new_value = value

    def save_edit(self, index: int) -> None:
        assert self.things[index].new_value is not None  # nosec
        self.things[index].value = str(self.things[index].new_value)  # str cast is because mypy can't figure out from the assert that it is not None.  # pylint: disable=line-too-long
        self.things[index].new_value = None
        self.things[index].is_editing = False

    def cancel_edit(self, index: int) -> None:
        self.things[index].new_value = None
        self.things[index].is_editing = False


def thing_box(thing: Thing, index: int) -> rx.Component:
    return rx.box(
        rx.cond(  # type: ignore
            # The issue: When this changes because of the the on_clicks below it
            # doesn't react, so you have to refresh the page to see the updated
            # state.
            thing.is_editing,
            rx.hstack(
                rx.debounce_input(
                    rx.input(
                        on_change=lambda value: State.update_new_value(index, value)  # type: ignore[call-arg,arg-type]  # pylint: disable=no-value-for-parameter
                    ),
                    value=thing.new_value,
                    debounce_timeout=250,
                ),
                rx.button(
                    rx.icon(tag="arrow_right"),
                    on_click=lambda: State.save_edit(index)  # type: ignore[attr-defined,call-arg,arg-type]  # pylint: disable=no-value-for-parameter,unknown-option-value
                ),
                rx.button(
                    rx.icon(tag="close"),
                    on_click=lambda: State.cancel_edit(index)  # type: ignore[attr-defined,call-arg,arg-type]  # pylint: disable=no-value-for-parameter,unknown-option-value
                ),
            ),
            rx.hstack(
                rx.box(
                    rx.markdown(
                        thing.value,
                    ),
                    background_color="#fafafa",
                    border_radius="10px",
                    padding="0.5em 1em 0.5em 1em",
                    width="100%",
                ),
                rx.button(
                    rx.icon(tag="edit"),
                    on_click=lambda: State.edit(index)  # type: ignore[attr-defined,call-arg,arg-type]  # pylint: disable=no-value-for-parameter,unknown-option-value
                )
            ),
        )
    )


def index() -> rx.Component:
    return rx.box(
        rx.vstack(
            rx.foreach(
                State.things,
                thing_box,
            ),
        ),
        align_items="center",
        display="flex",
        justify_content="center",
        min_height="100vh",
    )


app = rx.App() #state=State)
app.add_page(index)
app.compile()
```

When you run it up it looks like this. You should be able to edit any of the things, one or more at the same time.

![image](https://github.com/reflex-dev/reflex/assets/583807/1588f6b6-f242-494f-b8d4-42aa8b25f7ef)

It all works, except... you have to refresh the page after clicking any button so that the page state updates to reflect whatever the button was doing - updating a thing.is_editing to change what UI elements are visible and working.

What am I not getting? Cheers!

_________________________
Python: 3.11.4
Reflex: 0.2.6

**Reply by masenf** [reply, marked as answer]
The state cannot (yet) detect internal changes to a custom var (like `Thing`). You can do something like the following though, and yes, i realize that it's a bit ugly

```python
    def edit(self, index: int) -> None:
        thing = self.things[index]
        thing.new_value = thing.value
        thing.is_editing = True
        self.things[index] = thing
```

**The reassignment at the end is what will trigger the state to know that something in `State.things` has a delta to send to the frontend**

**Reply by masenf** [reply]
@nevdelap this actually got filed as a real issue yesterday https://github.com/reflex-dev/reflex/issues/1675

**Comment by nevdelap** [comment]
Thanks @masenf and @Lendemor.

Even just adding it like this works (as a workaround for now). :+1: 

```
    def edit(self, index: int) -> None:
        self.things[index].new_value = self.things[index].value
        self.things[index].is_editing = True
        self.things[index] = self.things[index] # <<<< this.
```
